### PR TITLE
Safely create VirtualCharSequence from string in StackFrameParser 

### LINKED ISSUE #59258: Assertion in StackFrameParser
Stack:

```
Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.EmbeddedLanguages.VirtualChars.VirtualCharSequence.StringChunk.this[int].get(int index) Line 112	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.EmbeddedLanguages.VirtualChars.VirtualCharSequence.this[int].get(int index) Line 70	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.EmbeddedLanguages.VirtualChars.VirtualCharSequence.Enumerator.Current.get() Line 21	C#
 	Microsoft.CodeAnalysis.Features.dll!Microsoft.CodeAnalysis.EmbeddedLanguages.StackFrame.StackFrameLexer.TryCreate(Microsoft.CodeAnalysis.EmbeddedLanguages.VirtualChars.VirtualCharSequence text) Line 49	C#
 	Microsoft.CodeAnalysis.Features.dll!Microsoft.CodeAnalysis.EmbeddedLanguages.StackFrame.StackFrameParser.TryParse(Microsoft.CodeAnalysis.EmbeddedLanguages.VirtualChars.VirtualCharSequence text) Line 49	C#
>	Microsoft.CodeAnalysis.Features.dll!Microsoft.CodeAnalysis.EmbeddedLanguages.StackFrame.StackFrameParser.TryParse(string text) Line 67	C#
 	Microsoft.CodeAnalysis.Features.dll!Microsoft.CodeAnalysis.StackTraceExplorer.DotnetStackFrameParser.TryParseLine(string line, out Microsoft.CodeAnalysis.StackTraceExplorer.ParsedFrame parsedFrame) Line 18	C#
 	Microsoft.CodeAnalysis.Features.dll!Microsoft.CodeAnalysis.StackTraceExplorer.StackTraceAnalyzer.Parse(string callstack, System.Threading.CancellationToken cancellationToken) Line 45	C#
```

Assertion is here:

```c#
#if DEBUG
                    // We should never have a properly paired high/low surrogate in a StringChunk. We are only created
                    // when the string has the same number of chars as there are VirtualChars.
                    if (char.IsHighSurrogate(_underlyingData[index]))
                    {
                        Debug.Assert(index + 1 >= _underlyingData.Length ||
                                     !char.IsLowSurrogate(_underlyingData[index + 1]));
                    }
#endif
```

The core issue is related to this line:

```c#
        public static StackFrameTree? TryParse(string text)
            => TryParse(VirtualCharSequence.Create(0, text));
```

It's not safe to blindly convert a string to a VirtualCharSequence.  That method is intended only for strings without escapes and withut surrogate chars in it.

### PR DESCRIPTION
**Pull request by ryzngard**
Fixes #59258

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/EmbeddedLanguages/VirtualChars/VirtualChar.cs [hunk 1]
```diff
@@ -67,6 +69,41 @@ public static VirtualChar Create(char surrogateChar, TextSpan span)
...
+        /// consumed by the <see cref="VirtualChar"/> that was added
+        /// </summary>
+        /// <param name="text">The whole text where index and offset</param>
+        /// <param name="index">The starting index within the string</param>
+        /// <param name="createSpan">Function to create the span for the VirtualChar, since it may be different than the index. Arguments are (start, length)</param>
+        /// <param name="consumedCharacters">Number of characters consumed. If iterating through the characters in a string, the index should be incremented by this amount.</param>
+        /// <returns></returns>
+        public static VirtualChar CreateNextInString(string text, int index, Func<int, int, TextSpan> createSpan, out int consumedCharacters)
+        {
+            if (Rune.TryCreate(text[index], out var rune))
```
Comment (ryzngard) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/EmbeddedLanguages/VirtualChars/VirtualChar.cs:85]: Ported from `AbstractVirtualCharService.cs`

File: src/Features/Core/Portable/EmbeddedLanguages/StackFrame/StackFrameLexer.cs [hunk 2]
```diff
@@ -23,38 +23,20 @@ internal struct StackFrameLexer
...
-        }
-
-        public static StackFrameLexer? TryCreate(VirtualCharSequence text)
+        public static StackFrameLexer? TryCreate(VirtualCharSequence virtualChars)
         {
-            foreach (var c in text)
+            foreach (var virtualChar in virtualChars)
             {
-                if (c.Value == '\r' || c.Value == '\n')
+                if (virtualChar.Value == '\r' || virtualChar.Value == '\n')
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/EmbeddedLanguages/StackFrame/StackFrameLexer.cs:33]:
```suggestion
                if (virtualChar.Value is '\r' or '\n')
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/EmbeddedLanguages/VirtualChars/AbstractVirtualCharService.cs [hunk 3]
```diff
@@ -159,25 +159,16 @@ private void CheckInvariants(SyntaxToken token, VirtualCharSequence result)
...
                 else
                 {
-                    // Something that couldn't be encoded as runes.
-                    Debug.Assert(char.IsSurrogate(tokenText[index]));
-                    result.Add(VirtualChar.Create(tokenText[index], new TextSpan(offset + index, 1)));
-                    index += 1;
+                    var virtualChar = VirtualChar.CreateNextInString(
+                        tokenText,
+                        index,
+                        (start, length) => new TextSpan(offset + index, length),
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/EmbeddedLanguages/VirtualChars/AbstractVirtualCharService.cs:167]: not happy about a lambda here.

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/EmbeddedLanguages/VirtualChars/AbstractVirtualCharService.cs:167]: this is called for every char in the string.

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/EmbeddedLanguages/VirtualChars/AbstractVirtualCharService.cs:167]: can you not just pass in the offset to CreateNextInString ?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/EmbeddedLanguages/VirtualChars/VirtualCharSequence.cs [hunk 4]
```diff
@@ -32,9 +32,32 @@ internal partial struct VirtualCharSequence
         public static VirtualCharSequence Create(ImmutableArray<VirtualChar> virtualChars)
             => new(new ImmutableArrayChunk(virtualChars));
 
-        public static VirtualCharSequence Create(int firstVirtualCharPosition, string underlyingData)
+        public static VirtualCharSequence UnsafeCreateFromAlreadyValidatedString(int firstVirtualCharPosition, string underlyingData)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/EmbeddedLanguages/VirtualChars/VirtualCharSequence.cs:35]: doc.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/EmbeddedLanguages/VirtualChars/VirtualCharSequence.cs [hunk 5]
```diff
@@ -32,9 +32,32 @@ internal partial struct VirtualCharSequence
...
             => new(new StringChunk(firstVirtualCharPosition, underlyingData));
 
+        /// <summary>
+        /// Takes in a string value and returns the <see cref="VirtualChar"/>s corresponding to
+        /// each char in the string using <see cref="VirtualChar.CreateNextInString(string, int, Func{int, int, TextSpan}, out int)"/>. Should be used over <see cref="VirtualCharSequence.UnsafeCreateFromAlreadyValidatedString(int, string)"/>
+        /// in cases where the string is arbitrary input (such as clipboard text).
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/EmbeddedLanguages/VirtualChars/VirtualCharSequence.cs:41]:
```suggestion
        /// Takes in a string value and returns the <see cref="VirtualChar"/>s corresponding to
        /// each char in the string using <see cref="VirtualChar.CreateNextInString(string, int, Func{int, int, TextSpan}, out int)"/>.
        /// Should be used over <see cref="VirtualCharSequence.UnsafeCreateFromAlreadyValidatedString(int, string)"/>
        /// in cases where the string is arbitrary input (such as clipboard text). 
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/EmbeddedLanguages/VirtualChars/VirtualCharSequence.cs [hunk 6]
```diff
@@ -32,9 +32,32 @@ internal partial struct VirtualCharSequence
...
+        public static VirtualCharSequence SafeCreateFromUnvalidatedString(string text)
+        {
+            using var _ = ArrayBuilder<VirtualChar>.GetInstance(out var builder);
+            for (var i = 0; i < text.Length; /* intentionally empty */)
+            {
+                var virtualChar = VirtualChar.CreateNextInString(
+                    text,
+                    i,
+                    static (start, length) => new TextSpan(start, length),
+                    out var consumedCharacters);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/EmbeddedLanguages/VirtualChars/VirtualCharSequence.cs:52]: consider reutrning a tuple.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/EmbeddedLanguages/VirtualChars/AbstractVirtualCharService.cs (hunk 3 above)
Comment (ryzngard) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/EmbeddedLanguages/VirtualChars/AbstractVirtualCharService.cs:167]: Sure, I'll change to take in an offset

File: src/EditorFeatures/Test/EmbeddedLanguages/StackFrame/StackFrameParserTests.cs [hunk 7]
```diff
@@ -422,6 +422,7 @@ public void TestFileInformation_InvalidDirectory()
         [InlineData("M.N(\r\n)")]
         [InlineData("M.N(\r)")]
         [InlineData("M.N(\n)")]
+        [InlineData("🤔")] // Surrogate pair for emoji. See https://github.com/dotnet/roslyn/issues/59258
```
[the remark this answers is not in the document]
Comment (ryzngard) [reply src/EditorFeatures/Test/EmbeddedLanguages/StackFrame/StackFrameParserTests.cs:425]: Can you use WorkItem with InlineData? I'll see if I can find one...

[the remark this answers is not in the document]
Comment (Youssef1313) [reply src/EditorFeatures/Test/EmbeddedLanguages/StackFrame/StackFrameParserTests.cs:425]: @ryzngard You can add it like:

```csharp
[InlineData("🤔"), WorkItem(....)]
```

The attribute itself is definitely tied to the whole test, but for the human eye, it's clear it's for this specific `InlineData`

**Review by JoeRobich — APPROVED** [review]

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/EmbeddedLanguages/VirtualChars/AbstractVirtualCharService.cs [hunk 8]
```diff
@@ -166,7 +166,16 @@ private void CheckInvariants(SyntaxToken token, VirtualCharSequence result)
...
+                    var (virtualChar, consumedCharacters) = VirtualChar.CreateNextInString(
+                        tokenText,
+                        index,
+                        offset);
+
+                    result.Add(virtualChar);
+                    index += consumedCharacters;
                 }
 
                 index += ConvertTextAtIndexToRune(tokenText, index, result, offset);
```
Comment (JoeRobich) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/EmbeddedLanguages/VirtualChars/AbstractVirtualCharService.cs:181]: This seems to already be handled by the `else` clause that you added. By having it here this code is executing twice and some of the text is not covered by the earlier quote and brace checks.

```suggestion
```

Comment (ryzngard) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/EmbeddedLanguages/VirtualChars/AbstractVirtualCharService.cs:181]: Not entirely sure I follow. If we move the characters by a certain amount the index will always need to be added to. `CreateNextInString` will consume some number of characters to produce a `virtualChar`. That's separate from `ConvertTextAtIndexToRune`

**Pull request closed without merging**